api/v1alpha1/nnf_port_manager_types.go

```diff
@@ -0,0 +1,136 @@
+
+// NnfPortManagerAllocationSpec defines the desired state for a single port allocation
+type NnfPortManagerAllocationSpec struct {
+	// Requester is an object reference to the requester of a ports.
+	Requester corev1.ObjectReference `json:"requester"`
+
+	// Count is the number of desired ports the requester needs. The port manager
+	// will attempt to allocate this many ports.
+	// +kubebuilder:default:=1
+	Count int `json:"count"`
+}
+
+// NnfPortManagerSpec defines the desired state of NnfPortManager
+type NnfPortManagerSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// SystemConfiguration is an object reference to the system configuration. The
+	// Port Manager will use the available ports defined in the system configuration.
+	SystemConfiguration corev1.ObjectReference `json:"systemConfiguration"`
+
+	// Allocations is a list of allocation requests that the Port Manager will attempt
+	// to satisfy. To request port resources from the port manager, clients should add
+	// an entry to the allocations. Entries must be unique. The port manager controller
+	// will attempt to allocate port resources for each allocation specification in the
+	// list. To remove an allocation and free up port resources, remove the allocation
+	// from the list.
+	Allocations []NnfPortManagerAllocationSpec `json:"allocations"`
+}
+
+// AllocationStatus is the current status of a port requestor. A port that is in use by the respective owner
+// will have a status of "InUse". A port that is freed by the owner but not yet reclaimed by the port manager
+// will have a status of "Free". Any other status value indicates a failure of the port allocation.
+// +kubebuilder:validation:Enum:=InUse;Free;InvalidConfiguration;InsufficientResources
+type NnfPortManagerAllocationStatusStatus string
+
+const (
+	NnfPortManagerAllocationStatusInUse                 NnfPortManagerAllocationStatusStatus = "InUse"
+	NnfPortManagerAllocationStatusFree                  NnfPortManagerAllocationStatusStatus = "Free"
+	NnfPortManagerAllocationStatusInvalidConfiguration  NnfPortManagerAllocationStatusStatus = "InvalidConfiguration"
+	NnfPortManagerAllocationStatusInsufficientResources NnfPortManagerAllocationStatusStatus = "InsufficientResources"
+	// NOTE: You must ensure any new value is added to the above kubebuilder validation enum
+)
+
+// NnfPortManagerAllocationStatus defines the allocation status of a port for a given requester.
+type NnfPortManagerAllocationStatus struct {
+	// Requester is an object reference to the requester of the port resource, if one exists, or
+	// empty otherwise.
+	Requester *corev1.ObjectReference `json:"requester,omitempty"`
+
+	// Ports is list of ports allocated to the owning resource.
+	Ports []uint16 `json:"ports,omitempty"`
+
+	// Status is the ownership status of the port.
+	Status NnfPortManagerAllocationStatusStatus `json:"status"`
```

Comment:
Would it be better to call this a state instead of having the double status?

Reply:
We use 'Status' everywhere, so I prefer it this way. If this were any other language I'd used a nested class.